core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/ShardedHttpExecutionServiceCluster.scala

```diff
@@ -54,22 +53,108 @@ class ShardedHttpExecutionServiceCluster (readMembers: Map[ExecutionServiceId,Ex
   def version: Future[ExecutionServiceVersion] =
     getRandomReadMember.version
 
+  // two possibilities here:
+  //
+  // (classic case) if the workflow is a top-level workflow of a submission, it has a row in the DB and an
+  // association with a specific execution service shard.  executionServiceKeyOpt = Some(shard)
+  //
+  // if it's a subworkflow (or sub-sub-workflow, etc) it's not present in the Rawls DB and we don't know which
+  // execution service shard has processed it.  executionServiceKeyOpt = None.  Query all execution service shards
+  // for the workflow to learn its submission association and which shard processed it.  In practice, one shard does
+  // everything except for some older workflows on shard 2.  Revisit this if that changes!
+
+  val SUBMISSION_ID_KEY = "workbench-submission-id"
+
+  private def parseSubWorkflowIdsFromMetadata(metadata: String): Seq[String] = {
+
+  //    Workflow metadata has this structure:
+  //
+  //    {
+  //      "calls" : {
+  //        "foo.bar" : [
+  //          {
+  //            "subWorkflowId" : "69581e76-2eb0-4179-99b9-958d210ebc4b", ...
+  //          }, ...
+  //        ], ...
+  //      },  ...
+  //    }
+
+
+    import spray.json._
+    import spray.json.DefaultJsonProtocol._
+
+    for {
+      callsObj <- JsonParser(metadata).asJsObject().getFields("calls")
+      call <- callsObj.asJsObject().fields.values
+      shard <- call.asInstanceOf[JsArray].elements
+      id <- shard.asJsObject().getFields("subWorkflowId")
+    } yield id.convertTo[String]
+  }
+
+  // parse subworkflow IDs from the parent workflow's metadata and label each of them with the Submission ID
+  // so they can be found by future workflow metadata queries
+
+  private def labelSubWorkflowsWithSubmissionId(submissionId: String, executionServiceId: ExecutionServiceId, parentWorkflowMetadata: JsObject, userInfo: UserInfo): Future[Seq[ExecutionServiceLabelResponse]] = {
+    // traverse = stop on "first" failure and propagate
+    Future.traverse(parseSubWorkflowIdsFromMetadata(parentWorkflowMetadata.compactPrint)) { subWorkflowId =>
```

Comment:
I find it icky that we are printing a `JsObject` to string here, then in `parseSubWorkflowIdsFromMetadata` we immediately parse it back to a `JsObject`. It looks like `parseSubWorkflowIdsFromMetadata` is not used anywhere else? Why not have it accept a `JsObject`?

Reply:
ew - must have missed that on some refactoring round.  Will fix.